lib/Dialect/FIRRTL/Transforms/LowerBundleVectorTypes.cpp

```diff
@@ -0,0 +1,967 @@
+  SmallVector<unsigned> argsToRemove;
+  // First get all the info for existing ports
+  SmallVector<std::pair<ModulePortInfo, SmallVector<NamedAttribute>>> newArgs;
+  for (auto port : llvm::enumerate(module.getPorts())) {
+    SmallVector<NamedAttribute> argAttrs;
+    AnnotationSet::forPort(module, port.index(), argAttrs);
+    newArgs.push_back(std::make_pair(port.value(), argAttrs));
+  }
+
+  for (size_t argIndex = 0; argIndex < newArgs.size(); ++argIndex) {
+    auto lowerings = lowerArg(module, argIndex, newArgs);
+    if (!lowerings.empty()) {
+      auto arg = module.getPortArgument(argIndex);
+      processUsers(arg, lowerings);
+      argsToRemove.push_back(argIndex);
+    }
+  // lowerArg might have invalidated any reference to newArgs, be careful
+  }
+
+  // Remove block args that have been lowered.
+  body->eraseArguments(argsToRemove);
+  for (auto ii = argsToRemove.rbegin(), ee = argsToRemove.rend(); ii != ee;
+       ++ii)
+    newArgs.erase(newArgs.begin() + *ii);
+
+  SmallVector<NamedAttribute, 8> newModuleAttrs;
+
+  // Copy over any attributes that weren't original argument attributes.
+  for (auto attr : module->getAttrDictionary())
+      // Drop old "portNames", directions, and argument attributes.  These are
+      // handled differently below.
+      if (attr.first != "portNames" && attr.first != direction::attrKey &&
+          attr.first != mlir::function_like_impl::getArgDictAttrName())
+        newModuleAttrs.push_back(attr);
+
+  SmallVector<Attribute> newArgNames;
+  SmallVector<Direction> newArgDirections;
+  SmallVector<Attribute, 8> newArgAttrs;
+  for (auto &port : newArgs) {
+    newArgNames.push_back(port.first.name);
+    newArgDirections.push_back(port.first.direction);
+    newArgAttrs.push_back(port.first.annotations.getArgumentAttrDict(port.second));
+  }
+  newModuleAttrs.push_back(NamedAttribute(Identifier::get("portNames", context),
+                                          builder->getArrayAttr(newArgNames)));
+  newModuleAttrs.push_back(
+      NamedAttribute(Identifier::get(direction::attrKey, context),
+                     direction::packAttribute(newArgDirections, context)));
+
+  // Attach new argument attributes.
+  newModuleAttrs.push_back(NamedAttribute(
+      builder->getIdentifier(mlir::function_like_impl::getArgDictAttrName()),
+      builder->getArrayAttr(newArgAttrs)));
+
+  // Update the module's attributes.
+  module->setAttrs(newModuleAttrs);
+
+  // Keep the module's type up-to-date.
+  auto moduleType = builder->getFunctionType(body->getArgumentTypes(), {});
+  module->setAttr(module.getTypeAttrName(), TypeAttr::get(moduleType));
+}
+
+/// Lower a wire op with a bundle to multiple non-bundled wires.
+void TypeLoweringVisitor::visitDecl(WireOp op) {
+  auto clone = [&](FlatBundleFieldEntry field, StringRef name,
+                   SmallVector<Attribute> &attrs) -> Operation * {
+    return builder->create<WireOp>(field.type, name, attrs);
+  };
+  lowerProducer(op, clone);
+}
```

Comment:
I love the similarity between all the `visitDecl`s!

Reply:
It was a nice cleanup.